src/control/mod.rs

```diff
@@ -1067,10 +1072,33 @@ impl std::fmt::Debug for Mode {
             .field("hskew", &self.hskew())
             .field("vscan", &self.vscan())
             .field("vrefresh", &self.vrefresh())
+            .field("type_", &self.type_())
             .finish()
     }
 }
 
+/// Display mode type flags
+#[repr(u32)]
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+pub enum ModeType {
```

Comment:
With these being bitflags they should probably be implemented using the `bitflags!` crate as an enum doesn't represent multiple variants at once.

At the same time isn't it easier to expose this publicly in `type_()`? Makes it more clear in the API what's used, gives easy access to things like `.contains()`, and still allows converting to the underlying raw `u32` with `.bits()`.

Reply:
drm-rs currently does not depend on `bitflags` and I am not sure we want to.

This style is inline with what we are doing today. Which is [using arrays of flags](https://github.com/Smithay/drm-rs/blob/develop/src/control/mod.rs#L761) *or* [just using integer types](https://github.com/Smithay/drm-rs/blob/develop/src/control/mod.rs#L359).

I do not think we have an example of returning bitflags in the library yet, but the `ModeType` as proposed still has value for testing through e.g. `mode.type_() & (ModeType::Preferred as u32)`.

This is a bit more cumbersome, but I think I am in favor of including it this way for now and discuss the usage of `bitflags` somewhere else. If we would start using that crate, we *should* use it consistently through-out the whole crate, which will then be a breaking change for a `0.6.0` version, while this can be shipped as `0.5.1`.

Reply:
> drm-rs currently does not depend on `bitflags` and I am not sure we want to.

I certainly hope this can be done at some point, it makes working with bitflags much easier than Rust enums and slices.

> This style is inline with what we are doing today. Which is [using arrays of flags](https://github.com/Smithay/drm-rs/blob/develop/src/control/mod.rs#L761) _or_ [just using integer types](https://github.com/Smithay/drm-rs/blob/develop/src/control/mod.rs#L359).
> 
> I do not think we have an example of returning bitflags in the library yet, but the `ModeType` as proposed still has value for testing through e.g. `mode.type_() & (ModeType::Preferred as u32)`.

Would look something like https://github.com/MarijnS95/drm-rs/commits/bitflags, but not applied to `ModeType` and others just yet. It feels weird to accept integer bitflags - which are supposed to be easily representable - as an array of integers holding the individual bits. I don't think `drm-rs` separates them out anywhere into a vec anyway, and a `u32` doesn't represent the possible return values in the way I would expect from a high-level API.

For this PR the least that could be added is:

```rust
/// Returns the [`ModeType`] bitmask of this mode
```

To make it clear what the possible values are - same for the (linked) uses of `flags: u32` if we don't end up going with bitflags.

> This is a bit more cumbersome, but I think I am in favor of including it this way for now and discuss the usage of `bitflags` somewhere else. If we would start using that crate, we _should_ use it consistently through-out the whole crate, which will then be a breaking change for a `0.6.0` version, while this can be shipped as `0.5.1`.

Fully agreed. It doesn't make sense to introduce that in this PR while nothing else uses it just yet. If you like what you see above I can possibly set up a WIP PR and convert a few more occurrences. Not sure what to do with `DRM_MODE_PAGE_FLIP_TARGET` just yet, though.

Note that quite a few crates in our dependency graph use `bitflags` already.

Reply:
Anything still has to happen here? Looking at `DRM_MODE_PAGE_FLIP_TARGET` it seems either `DRM_MODE_PAGE_FLIP_TARGET_ABSOLUTE` or `DRM_MODE_PAGE_FLIP_TARGET_RELATIVE` should be passed in the _`flags`_ field, and the `reserved`/`sequence` field should be used for an integer denoting the desired vblank instead (instead of the flag). We might want to fix that.